### PR TITLE
feat: upgraded the app with scheduler module to SDK v0.50

### PR DESCRIPTION
**Pull request by deepan95dev**

### CODE REVIEW

File: x/scheduler/keeper/wasm_handler_test.go [hunk 1]
```diff
@@ -1,136 +1,137 @@
...
-
-	for _, tt := range tests {
-		t.Run(tt.name, func(t *testing.T) {
-			testKeeper := keeper.Keeper{}
-			actual, _ := testKeeper.UnmarshallJob(tt.input)
-			asserter.Equal(string(tt.expected.Payload), string(actual.Payload))
-		})
-	}
-}
+//
```
Comment (byte-bandit) [inline x/scheduler/keeper/wasm_handler_test.go:3]: This will come back later?

File: x/scheduler/module.go [hunk 2]
```diff
@@ -99,6 +99,14 @@ type AppModule struct {
 	bankKeeper    types.BankKeeper
 }
 
+func (am AppModule) IsOnePerModuleType() {
+	//TODO implement me
```
Comment (byte-bandit) [inline x/scheduler/module.go:103]: When is this coming?

File: x/scheduler/types/keeper.go [hunk 3]
```diff
@@ -1,18 +1,19 @@
...
 //go:generate mockery --name=Keeper
 type Keeper interface {
 	AddNewJob(ctx sdk.Context, job *Job) error
-	GetAccount(ctx sdk.Context, addr sdk.AccAddress) authtypes.AccountI
+	GetAccount(ctx context.Context, addr sdk.AccAddress) types.AccountI
 	GetJob(ctx sdk.Context, jobID string) (*Job, error)
-	Logger(ctx sdk.Context) log.Logger
-	PreJobExecution(ctx sdk.Context, job *Job) error
+	Logger(ctx context.Context) log.Logger
+	PreJobExecution(ctx context.Context, job *Job) error
```
Comment (byte-bandit) [inline x/scheduler/types/keeper.go:16]: Some of the keeper functions are using `sdk.Context`, now some others are using `context.Context`. Where's the difference coming from and can we align on just using one of them?

**Review by byte-bandit — CHANGES_REQUESTED** [review]

File: x/scheduler/keeper/wasm_handler_test.go (hunk 1 above)
Comment (deepan95dev) [reply x/scheduler/keeper/wasm_handler_test.go:3]: yes, test files require some of the other modules. After the merging the modular PRs, will enable these.

File: x/scheduler/module.go (hunk 2 above)
Comment (deepan95dev) [reply x/scheduler/module.go:103]: There is no usage for this function but required by sdk keeper struct.

[Ref](https://github.com/cosmos/cosmos-sdk/blob/7d5c2dbae68530bc9070983ce0b2c8a1efe23412/x/bank/module.go#L100) usage in sdk bank module

File: x/scheduler/types/keeper.go (hunk 3 above)
Comment (deepan95dev) [reply x/scheduler/types/keeper.go:16]: We replaced the structs defined within paloma modules and types to `context.Context` we did not replace the ones which were defined by cosmos-sdk struct. [Here](https://github.com/cosmos/cosmos-sdk/blob/7d5c2dbae68530bc9070983ce0b2c8a1efe23412/x/group/keeper/keeper.go#L238) is the sample keeper from cosmos-sdk which still uses sdk.context

Comment (byte-bandit) [reply x/scheduler/types/keeper.go:16]: What's the idea behind this? Why not use `sdk.Context` for custom handlers as well?

File: x/scheduler/module.go (hunk 2 above)
Comment (byte-bandit) [reply x/scheduler/module.go:103]: In that case, remove the TODO and replace with the contents of your comment, including the reference link. Thanks.

Comment (deepan95dev) [reply x/scheduler/module.go]: Done

File: x/scheduler/types/keeper.go (hunk 3 above)
Comment (deepan95dev) [reply x/scheduler/types/keeper.go]: Sure, I just checked there were few ununsed types which were using `sdk.context` now updated to `context.Context`

**Comment by aleem1314** [comment]
Check #1036 PR review comments. Apply relevant changes here too.

**Comment by deepan95dev** [comment]
> Check #1036 PR review comments. Apply relevant changes here too.

done

File: proto/palomachain/paloma/scheduler/tx.proto [hunk 4]
```diff
@@ -3,16 +3,32 @@ package palomachain.paloma.scheduler;
...
+
+  // authority is the address of the governance account.
+  string authority = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+
+  // params defines the x/scheduler parameters to update.
+  Params params = 2 [(gogoproto.nullable) = false];
+}
+
+message MsgUpdateParamsResponse {}
+
```
Comment (byte-bandit) [inline proto/palomachain/paloma/scheduler/tx.proto:31]: Do we really need this now? Paloma isn't really using params yet.

If so, please make sure you add the `Metadata` field that all other Paloma messages are embedding.

File: x/scheduler/keeper/keeper.go [hunk 5]
```diff
@@ -21,80 +22,78 @@ import (
...
-func (k Keeper) GetAccount(ctx sdk.Context, addr sdk.AccAddress) authtypes.AccountI {
+func (k Keeper) GetAccount(ctx context.Context, addr sdk.AccAddress) authtypes.AccountI {
 	return k.account.GetAccount(ctx, addr)
 }
 
-func (k Keeper) Logger(ctx sdk.Context) log.Logger {
-	return ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
+func (k Keeper) Logger(ctx context.Context) log.Logger {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	return sdkCtx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
```
Comment (byte-bandit) [inline x/scheduler/keeper/keeper.go:73]: Use `liblog` for logging.

Comment (byte-bandit) [reply x/scheduler/keeper/keeper.go:73]: Everywhere, exclusively.  No other logging calls in production code.

File: x/scheduler/keeper/msg_server.go [hunk 6]
```diff
@@ -15,3 +21,21 @@ func NewMsgServerImpl(keeper types.Keeper) types.MsgServer {
...
+	if err := ms.SetParams(ctx, req.Params); err != nil {
+		return nil, err
+	}
+
+	return &types.MsgUpdateParamsResponse{}, nil
+}
+
+func (ms msgServer) SetParams(ctx context.Context, params types.Params) error {
+	ms.SetParams(ctx, params)
+	return nil
```
Comment (byte-bandit) [inline x/scheduler/keeper/msg_server.go:40]: Does the `0.50` upgrade enforce this?

File: x/scheduler/types/msgs.go [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+// GetSignBytes implements the LegacyMsg interface.
+func (m MsgUpdateParams) GetSignBytes() []byte {
+	return sdk.MustSortJSON(ModuleCdc.MustMarshalJSON(&m))
+}
+
+// GetSigners returns the expected signers for a MsgUpdateParams message.
+func (m *MsgUpdateParams) GetSigners() []sdk.AccAddress {
+	addr, _ := sdk.AccAddressFromBech32(m.Authority)
+	return []sdk.AccAddress{addr}
```
Comment (byte-bandit) [inline x/scheduler/types/msgs.go:18]: Paloma is using a field called `Metadata` for all custom messages. Please take a look at other messages to see how it's used and adapt this to match.

**Review by byte-bandit — CHANGES_REQUESTED** [review]

File: x/scheduler/keeper/keeper.go (hunk 5 above)
Comment (deepan95dev) [reply x/scheduler/keeper/keeper.go:73]: Did you mean to replace the usage of fmt here? It was there in the master branch already [here](https://github.com/palomachain/paloma/blob/72fcb671899c8eb0557e5342cd7547e9655e14c5/x/scheduler/keeper/keeper.go#L75)

File: x/scheduler/keeper/msg_server.go (hunk 6 above)
Comment (deepan95dev) [reply x/scheduler/keeper/msg_server.go:40]: let me check the usage of this function in the scheduler module and update

File: x/scheduler/keeper/keeper.go (hunk 5 above)
Comment (byte-bandit) [reply x/scheduler/keeper/keeper.go:73]:
```suggestion
		return liblog.FromSDKLogger(ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName)))
```

**Review by aleem1314 — CHANGES_REQUESTED** [review]
Still there are few things need to be updated.
Check https://github.com/palomachain/paloma/pull/1036#pullrequestreview-1763965986

File: x/scheduler/keeper/msg_server.go (hunk 6 above)
Comment (deepan95dev) [reply x/scheduler/keeper/msg_server.go:40]: sorry for the delay, just verified that above functions are required and used in module.go https://github.com/palomachain/paloma/blob/469ba6bf4ecc59010d7c25ce35ce2162ff49c8ac/x/scheduler/module.go#L133

File: x/scheduler/types/msgs.go (hunk 7 above)
Comment (deepan95dev) [reply x/scheduler/types/msgs.go]: Updated to default and pushed.

File: x/scheduler/keeper/keeper.go (hunk 5 above)
Comment (deepan95dev) [reply x/scheduler/keeper/keeper.go]: Updated and pushed

File: proto/palomachain/paloma/scheduler/tx.proto (hunk 4 above)
Comment (deepan95dev) [reply proto/palomachain/paloma/scheduler/tx.proto]: removed and pushed

File: x/scheduler/keeper/msg_server.go (hunk 6 above)
Comment (deepan95dev) [reply x/scheduler/keeper/msg_server.go]: removed and pushed

File: x/scheduler/types/msgs.go [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+var _ sdk.Msg = &MsgUpdateParams{}
+
+// GetSignBytes implements the LegacyMsg interface.
+func (m MsgUpdateParams) GetSignBytes() []byte {
+	return sdk.MustSortJSON(ModuleCdc.MustMarshalJSON(&m))
+}
+
+// GetSigners returns the expected signers for a MsgUpdateParams message.
+func (m *MsgUpdateParams) GetSigners() []sdk.AccAddress {
+	return []sdk.AccAddress{}
```
Comment (byte-bandit) [inline x/scheduler/types/msgs.go:17]: Please take a look at how this interface is implemented for other messages in Paloma and update this code to match.

File: x/scheduler/types/msgs.go [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+	if _, err := sdk.AccAddressFromBech32(m.Authority); err != nil {
+		return sdkerrors.Wrap(err, "invalid authority address")
+	}
+
+	if err := m.Params.Validate(); err != nil {
+		return err
+	}
+
+	return nil
+}
```
Comment (byte-bandit) [inline x/scheduler/types/msgs.go:31]: Also this.

**Review by byte-bandit — CHANGES_REQUESTED** [review]

File: app/app.go [hunk 10]
```diff
@@ -598,17 +601,17 @@ func New(
...
-	// 	[]xchain.Bridge{
-	// 		app.EvmKeeper,
-	// 	},
-	// )
+	app.SchedulerKeeper = *schedulermodulekeeper.NewKeeper(
+		appCodec,
+		runtime.NewKVStoreService(keys[evidencetypes.StoreKey]),
+		memKeys[schedulermoduletypes.MemStoreKey],
+		app.AccountKeeper,
+		nil,
```
Comment (aleem1314) [inline app/app.go:609]: EVM module is enabled. You should pass EVM module keeper here.

File: app/app.go [hunk 11]
```diff
@@ -598,17 +601,17 @@ func New(
...
+		[]xchain.Bridge{
+			//app.EvmKeeper,
```
Comment (aleem1314) [inline app/app.go:611]:
```suggestion
			app.EvmKeeper,
```

File: x/scheduler/keeper/exported/exported.go [hunk 12]
```diff
@@ -0,0 +1,20 @@
+package exported
```
Comment (aleem1314) [inline x/scheduler/keeper/exported/exported.go:1]: This file is not required. You can delete this.

File: x/scheduler/keeper/keeper.go [hunk 13]
```diff
@@ -21,80 +22,78 @@ import (
 )
 
 type Keeper struct {
-	cdc        codec.BinaryCodec
-	storeKey   storetypes.StoreKey
-	memKey     storetypes.StoreKey
-	paramstore paramtypes.Subspace
+	cdc      codec.BinaryCodec
+	storeKey store.KVStoreService
+	memKey   storetypes.StoreKey
```
Comment (aleem1314) [inline x/scheduler/keeper/keeper.go:27]: You should use same type for both `storeKey` and `memKey`.

File: x/scheduler/keeper/keeper.go [hunk 14]
```diff
@@ -21,80 +22,78 @@ import (
...
-		account:    account,
-		EvmKeeper:  evmKeeper,
-		Chains:     cm,
+		cdc:       cdc,
+		storeKey:  storeKey,
+		memKey:    memKey,
+		account:   account,
+		EvmKeeper: evmKeeper,
+		Chains:    cm,
+		authority: authority,
```
Comment (aleem1314) [inline x/scheduler/keeper/keeper.go:59]: Where are we using `authority`?

File: x/scheduler/keeper/keeper.go [hunk 15]
```diff
@@ -21,80 +22,78 @@ import (
...
 
-func (k Keeper) PreJobExecution(ctx sdk.Context, job *types.Job) error {
+func (k Keeper) PreJobExecution(ctx context.Context, job *types.Job) error {
 	return k.EvmKeeper.PreJobExecution(ctx, job)
 }
 
 // store returns default store for this keeper!
-func (k Keeper) Store(ctx sdk.Context) sdk.KVStore {
-	return ctx.KVStore(k.storeKey)
+func (k Keeper) Store(ctx context.Context) storetypes.KVStore {
```
Comment (aleem1314) [inline x/scheduler/keeper/keeper.go:85]: This is not correct. You are returning `memStore`, it should be kvstore.

File: x/scheduler/keeper/keeper.go [hunk 16]
```diff
@@ -21,80 +22,78 @@ import (
...
 
-func (k Keeper) jobsStore(ctx sdk.Context) sdk.KVStore {
+func (k Keeper) jobsStore(ctx context.Context) storetypes.KVStore {
 	return prefix.NewStore(k.Store(ctx), types.KeyPrefix("jobs"))
 }
 
-func (k Keeper) AddNewJob(ctx sdk.Context, job *types.Job) error {
+func (k Keeper) AddNewJob(ctx context.Context, job *types.Job) error {
 	if k.JobIDExists(ctx, job.GetID()) {
-		return types.ErrJobWithIDAlreadyExists.Wrap(job.GetID())
```
Comment (aleem1314) [inline x/scheduler/keeper/keeper.go:97]: Why this change?

File: x/scheduler/types/msgs.go [hunk 17]
```diff
@@ -0,0 +1,31 @@
+package types
+
+import (
+	sdkerrors "cosmossdk.io/errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+var _ sdk.Msg = &MsgUpdateParams{}
```
Comment (aleem1314) [inline x/scheduler/types/msgs.go:8]: The `schedular` module does not have any params. You can remove this file.

File: app/app.go (hunk 10 above)
Comment (deepan95dev) [reply app/app.go]: enabled and pushed

File: app/app.go (hunk 11 above)
Comment (deepan95dev) [reply app/app.go]: done

File: third_party/cosmos/msg/v1/msg.proto [hunk 18]
```diff
@@ -0,0 +1,30 @@
+syntax = "proto3";
```
[the remark this answers is not in the document]
Comment (deepan95dev) [reply third_party/cosmos/msg/v1/msg.proto]: deleted

File: x/scheduler/keeper/exported/exported.go (hunk 12 above)
Comment (deepan95dev) [reply x/scheduler/keeper/exported/exported.go]: deleted

File: x/scheduler/types/msgs.go (hunk 9 above)
Comment (deepan95dev) [reply x/scheduler/types/msgs.go]: deleted this file for scheduler module

File: x/scheduler/types/msgs.go (hunk 8 above)
Comment (deepan95dev) [reply x/scheduler/types/msgs.go]: deleted this file for scheduler module

File: proto/palomachain/paloma/scheduler/tx.proto [hunk 19]
```diff
@@ -3,6 +3,7 @@ package palomachain.paloma.scheduler;
 
 import "palomachain/paloma/scheduler/job.proto";
 import "palomachain/paloma/valset/common.proto";
+import "palomachain/paloma/scheduler/params.proto";
```
Comment (byte-bandit) [inline proto/palomachain/paloma/scheduler/tx.proto:6]: Unneeded import?

File: x/scheduler/keeper/keeper.go (hunk 13 above)
Comment (deepan95dev) [reply x/scheduler/keeper/keeper.go]: Updated and pushed

File: x/scheduler/keeper/keeper.go (hunk 14 above)
Comment (deepan95dev) [reply x/scheduler/keeper/keeper.go]: removed and pushed

File: x/scheduler/keeper/keeper.go (hunk 16 above)
Comment (deepan95dev) [reply x/scheduler/keeper/keeper.go:97]: reverted and pushed

File: x/scheduler/types/msgs.go (hunk 17 above)
Comment (deepan95dev) [reply x/scheduler/types/msgs.go]: this file was deleted.

File: proto/palomachain/paloma/scheduler/tx.proto (hunk 19 above)
Comment (deepan95dev) [reply proto/palomachain/paloma/scheduler/tx.proto]: updated and pushed

File: x/scheduler/types/tx.pb.go [hunk 20]
```diff
@@ -29,6 +31,96 @@ var _ = math.Inf
 // proto package needs to be updated.
 const _ = proto.GoGoProtoPackageIsVersion3 // please upgrade the proto package
 
+type MsgUpdateParams struct {
```
[the remark this answers is not in the document]
Comment (deepan95dev) [reply x/scheduler/types/tx.pb.go]: generated the new proto files and pushed

File: x/scheduler/keeper/keeper.go (hunk 15 above)
Comment (deepan95dev) [reply x/scheduler/keeper/keeper.go:85]: I pushed above fix with reference link with comment [here](https://github.com/palomachain/paloma/pull/1044/commits/7f8e981e399c3b70a0631685787434721c9dd1ae) Once verify if its correct usage @aleem1314  @byte-bandit

File: x/scheduler/types/mocks/BankKeeper.go [hunk 21]
```diff
@@ -0,0 +1,50 @@
+// Code generated by mockery v2.38.0. DO NOT EDIT.
```
Comment (byte-bandit) [inline x/scheduler/types/mocks/BankKeeper.go:1]: Why are these new? Are they used anywhere?

**Review by byte-bandit — COMMENTED** [review]
LGTM. Please clean up unused files and make sure the linters run successfully. 

Revisit your commit history to stash and fixup where useful. Please use [convential commits](https://www.conventionalcommits.org/en/v1.0.0/) and make sure you [sign your commits](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits).

File: x/scheduler/types/mocks/BankKeeper.go (hunk 21 above)
Comment (deepan95dev) [reply x/scheduler/types/mocks/BankKeeper.go]: deleted unused mock files and pushed.

**Comment by deepan95dev** [comment]
> LGTM. Please clean up unused files and make sure the linters run successfully.
> 
> Revisit your commit history to stash and fixup where useful. Please use [convential commits](https://www.conventionalcommits.org/en/v1.0.0/) and make sure you [sign your commits](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits).

noted. Will follow the commit conventions and verified signatures for further commits.

**Review by byte-bandit — DISMISSED** [review]
LGTM, final review to @aleem1314

File: x/scheduler/types/params_legacy.go [hunk 22]
```diff
@@ -0,0 +1 @@
+package types
```
Comment (aleem1314) [inline x/scheduler/types/params_legacy.go:1]: You can delete this file.

**Review by aleem1314 — APPROVED** [review]
Left one comment. Pre approving

File: x/scheduler/types/params_legacy.go (hunk 22 above)
Comment (deepan95dev) [reply x/scheduler/types/params_legacy.go]: done

**Comment by deepan95dev** [comment]
@byte-bandit rebase and fixup done.

File: x/scheduler/types/params_legacy.go (hunk 22 above)
Comment (taariq) [reply x/scheduler/types/params_legacy.go]: Resolving.

File: x/scheduler/types/mocks/BankKeeper.go (hunk 21 above)
Comment (taariq) [reply x/scheduler/types/mocks/BankKeeper.go]: Resolved.

**Review by taariq — APPROVED** [review]
@byte-bandit @verabehr approving for merge.

**Comment by taariq** [comment]
@aleem1314 requires your approval here on review.

**Review by aleem1314 — APPROVED** [review]

**Pull request merged**